File: Bots/JavaScript/Skills/CodeFirst/WaterfallSkillBot/dialogs/update/updateDialog.js
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

const { InputHints, MessageFactory } = require('botbuilder');
const { ChoiceFactory, ChoicePrompt, ComponentDialog, ListStyle, WaterfallDialog, DialogTurnStatus } = require('botbuilder-dialogs');
const { Channels } = require('botbuilder-core');

const WATERFALL_DIALOG = 'WaterfallDialog';
const UPDATE_UNSUPPORTED = new Set([Channels.Emulator, Channels.Facebook, Channels.Webchat]);
const CHOICE_PROMPT = "ChoicePrompt";


class UpdateDialog extends ComponentDialog {
    /**
     * @param {string} dialogId
     */
    constructor(dialogId) {
        super(dialogId);

        this.updateTracker = {};

        this.addDialog(new ChoicePrompt(CHOICE_PROMPT));
        this.addDialog(new WaterfallDialog(WATERFALL_DIALOG, [
                this.HandleUpdateDialog.bind(this),
                this.FinalStepAsync.bind(this)
            ]));

        this.initialDialogId = WATERFALL_DIALOG;
    }

    /**
     * @param {import('botbuilder-dialogs').WaterfallStepContext} stepContext
     */
    async HandleUpdateDialog(stepContext) {
        let channel = stepContext.context.activity.channelId;

        if (UpdateDialog.isUpdateSupported(channel)){
            const conversationId = stepContext.context.activity.conversation.id;

            if (conversationId in this.updateTracker) {
                var tuple = this.updateTracker[conversationId];
                var activity = MessageFactory.text(`This message has been updated ${tuple[1]} time(s).`)
                activity.id = tuple[0];
                tuple[1]++;
                this.updateTracker[conversationId] = tuple;
                await stepContext.context.updateActivity(activity);
            }
            else
            {
                var id = await stepContext.context.sendActivity(MessageFactory.text("Here is the original activity"));
                this.updateTracker[conversationId] = [id.id, 1];
            }
        }
        else
        {
            await stepContext.context.sendActivity(MessageFactory.text(`Update is not supported in the ${channel} channel.`))
        }

        const messageText = 'Do you want to update the activity again?';
        const repromptMessageText = 'Please select a valid option';
        const options = {
            prompt: MessageFactory.text(messageText, messageText, InputHints.ExpectingInput),
            retryPrompt: MessageFactory.text(repromptMessageText, repromptMessageText, InputHints.ExpectingInput),
            choices: ChoiceFactory.toChoices(["yes", "no"]),
            style: ListStyle.list
        };

        // Ask the user to enter a card choice.
        return stepContext.prompt(CHOICE_PROMPT, options);
    }

    async FinalStepAsync(stepContext){
        const choice = stepContext.result.value;

        if (choice == "yes"){
            return stepContext.replaceDialog(this.initialDialogId);
        }
        
        this.updateTracker = {};
        return { status: DialogTurnStatus.complete };
    }

    static isUpdateSupported(channel) {
        return !UPDATE_UNSUPPORTED.has(channel);
    }
}

module.exports.UpdateDialog = UpdateDialog

File: Bots/DotNet/Skills/CodeFirst/TeamsSkillBot/Dialogs/Update/UpdateDialog.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Bot.Builder;
using Microsoft.Bot.Builder.Dialogs;
using Microsoft.Bot.Builder.Dialogs.Choices;
using Microsoft.Bot.Connector;

namespace Microsoft.BotFrameworkFunctionalTests.TeamsSkillBot.Dialogs.Update
{
    public class UpdateDialog : ComponentDialog
    {
        private Dictionary<string, (string, int)> _updateTracker;
        private readonly HashSet<string> _updateSupported = new HashSet<string>
        {
            Channels.Msteams,
            Channels.Slack,
            Channels.Telegram
        };

        public UpdateDialog()
             : base(nameof(UpdateDialog))
        {
            _updateTracker = new Dictionary<string, (string, int)>();
            AddDialog(new ChoicePrompt("ChoicePrompt"));
            AddDialog(new WaterfallDialog(nameof(WaterfallDialog), new WaterfallStep[] { HandleUpdateDialog, FinalStepAsync }));
            InitialDialogId = nameof(WaterfallDialog);
        }

        private async Task<DialogTurnResult> HandleUpdateDialog(WaterfallStepContext stepContext, CancellationToken cancellationToken)
        {
            var channel = stepContext.Context.Activity.ChannelId;
            if (_updateSupported.Contains(channel))
            {
                if (_updateTracker.ContainsKey(stepContext.Context.Activity.Conversation.Id))
                {
                    var conversationId = stepContext.Context.Activity.Conversation.Id;
                    (string, int) tuple = _updateTracker[conversationId];
                    var activity = MessageFactory.Text($"This message has been updated {tuple.Item2} time(s).");
                    tuple.Item2 += 1;
                    activity.Id = tuple.Item1;
                    _updateTracker[conversationId] = tuple;
                    await stepContext.Context.UpdateActivityAsync(activity, cancellationToken);
                }
                else
                {
                    var id = await stepContext.Context.SendActivityAsync(MessageFactory.Text("Here is the original activity"), cancellationToken);
                    _updateTracker.Add(stepContext.Context.Activity.Conversation.Id, (id.Id, 1));
                }
            }
            else
            {
                await stepContext.Context.SendActivityAsync(MessageFactory.Text($"Delete is not supported in the {channel} channel."), cancellationToken);
                return new DialogTurnResult(DialogTurnStatus.Complete);
            }

            // Create the PromptOptions from the skill configuration which contain the list of configured skills.
            var messageText = "Do you want to update the activity again?";
            var repromptMessageText = "Please select a valid answer";
            var options = new PromptOptions
            {
                Prompt = MessageFactory.Text(messageText, messageText),
                RetryPrompt = MessageFactory.Text(repromptMessageText, repromptMessageText),
                Choices = new List<Choice> { new Choice("Yes"), new Choice("No") },
                Style = ListStyle.List
            };

            // Ask the user to enter their name.
            return await stepContext.PromptAsync("ChoicePrompt", options, cancellationToken);
        }

        private async Task<DialogTurnResult> FinalStepAsync(WaterfallStepContext stepContext, CancellationToken cancellationToken)
        {
            var choice = ((FoundChoice)stepContext.Result).Value.ToLowerInvariant();

            if (choice == "yes")
            {
                return await stepContext.ReplaceDialogAsync(InitialDialogId, "Do you want to update the activity?", cancellationToken);
            }
            else
            {
                _updateTracker.Remove(stepContext.Context.Activity.Conversation.Id);
                return new DialogTurnResult(DialogTurnStatus.Complete);
            }
        }
    }
}
